auth/authentication.go

```diff
@@ -17,119 +16,163 @@ type Database interface {
 	GetApplicationByToken(token string) (*model.Application, error)
 	GetClientByToken(token string) (*model.Client, error)
 	GetPluginConfByToken(token string) (*model.PluginConf, error)
+
 	GetUserByName(name string) (*model.User, error)
 	GetUserByID(id uint) (*model.User, error)
+	CreateUser(user *model.User) error
+}
+
+// AuthenticationProvider provides authentication methods
+type AuthenticationProvider interface {
+	Authenticate(req *http.Request) (user *model.User, err error)
 }
 
 // Auth is the provider for authentication middleware
 type Auth struct {
-	DB Database
+	authenticators map[string]AuthenticationProvider
+	DB             Database
 }
 
-type authenticate func(tokenID string, user *model.User) (authenticated bool, success bool, userId uint, err error)
+// RegisterAuthenticationProvider registers a new authentication provider
+// use empty string as key to register internal credential manager
+func (a *Auth) RegisterAuthenticationProvider(key string, auth AuthenticationProvider) {
```

Comment:
Can RegisterAuthenticationProvider be called during runtime or must it be called at startup? In the first case we should protect `authenticators` with a mutex

Reply:
I think all providers should be registered during initialization so no concurrent write here :)